src/inmanta/compiler/__init__.py

```diff
@@ -314,3 +318,44 @@ def add_trace(exception: CompilerException) -> bool:
         add_trace(exception)
         exception.attach_compile_info(self)
         raise exception
+
+
+class Finalizers:
+    """
+    This class keeps all the finalizers that need to be called right after the compilation finishes
+    """
+
+    __finalizers: List[Callable[[], None]] = []
+
+    @classmethod
+    def add_function(cls, fnc: Callable[[], None]) -> None:
+        cls.__finalizers.append(fnc)
+
+    @classmethod
+    def call_finalizers(cls, should_log: bool = False) -> None:
+        """
+        by default this function will raise exceptions caused by errors in the finalizer functions
+        if 'should_log' is set to True the exceptions will not be raised but logged instead.
+        """
+        try:
+            for fnc in cls.__finalizers:
+                fnc()
+        except Exception as e:
+            if should_log:
+                LOGGER.error(f"Finalizers failed: {e}")
+            else:
+                raise
```

Comment:
Just a suggestion for brevity:
```suggestion
        except (Exception if should_log else ()) as e:
            LOGGER.error(f"Finalizers failed: {e}")
```